clarifies mf timespine & creates mini guide (#6836)
dbt-labs/docs.getdbt.com: website/docs/guides/mf-time-spine.md

    columns:
      - name: date_day
        description: The base date column for daily granularity
        granularity: day
```
</File>

This time spine YAML file:
- Defines `date_day` as the base column for daily granularity.
- Configures `time_spine` properties so MetricFlow can use the table

### Using an existing dim_date table

This optional approach reuses an existing table, saving you the effort of creating a new one. However if you created a time spine table from scratch, you can skip this section. 

If your project already includes a `dim_date` or similar table, you can configure it as a time spine:

1. Locate the existing table (`dim_date`).
2. Update `_models.yml` file to configure it as a time spine:

<File name='_models.yml'>

```yaml
models:
  - name: dim_date
    description: An existing date dimension table used as a time spine.
    time_spine:
      standard_granularity_column: date_day
    columns:
      - name: date_day
        granularity: day
```
</File>

This time spine YAML file configures the `time_spine` property so MetricFlow can use the table.

## Run and test the time spine

For the time spine table you created, let's run it and validate the output.

> 1. Run the time spine model to create the table and validate the output:

Author: ```suggestion
1. Run the following command:
```

   ```bash
   dbt run --select time_spine_daily
   ```

2. Check that the table:
   - Contains one row per day.
   - Covers the date range you want (5 years back to 30 days forward)

3. (Optional) If you have [metrics](/docs/build/metrics-overview) already defined in your project, you can query the table/metrics using [Semantic Layer commands](/docs/build/metricflow-commands) to validate the time spine. 
   
   Let's say you have a `revenue` metric defined. You can query the table/metrics using the following command:

    ```bash
    dbt sl query --metrics revenue --group-by metric_time
    ```

    This will output results similar to the following in the dbt Cloud IDE:

>     <Lightbox src="/img/quickstarts/dbt-cloud/validate-mf-timespine-output.jpg" title="Validate the time spine output in dbt Cloud IDE" />

Author: ```suggestion
    <Lightbox src="/img/quickstarts/dbt-cloud/validate-mf-timespine-output.jpg" title="Validate the metrics and time spine output in dbt Cloud IDE" />
```

---
title: "Quickstart with MetricFlow time spine"
id: "mf-time-spine"
level: 'Intermediate'
icon: 'guides'
tags: ['Quickstart']
hide_table_of_contents: true
recently_updated: true
---

<div style={{maxWidth: '900px'}}>

## Introduction

This guide explains how to configure a time spine using the [dbt Semantic Layer Jaffle shop example project](https://github.com/dbt-labs/jaffle-sl-template) as a reference. 

> ### What is a time spine?

Author: ```suggestion
### What is a time spine table?
```

A [time spine](/docs/build/metricflow-time-spine) is essential for time-based joins and aggregations in MetricFlow, the engine that powers the dbt Semantic Layer. It can be created from scratch or configured using an existing table (like a `dim_date` table).

To use MetricFlow with time-based metrics and dimensions, you must provide a time spine table. This table serves as the foundation for time-based joins and aggregations. You can either:

- Create a time spine SQL table from scratch, or
- Use an existing table in your project, like a `dim_date` table

And once you have a time spine table, you need to configure it in YAML to tell MetricFlow how to use it. This guide will show you how to do both!

### Prerequisites
Before you start, make sure you have:

> - A dbt project set up. If you don't have one, follow the [Quickstart guides](/guides?tags=Quickstart) guide to help you get started.

Author: ```suggestion
- A dbt project set up. If you don't have one, follow the [Semantic Layer quickstart guide](/guides/sl-snowflake-qs?step=1) or the [dbt Cloud quickstart guides)(/guides?tags=Quickstart) guide to help you get started.
```

            "DATE('2030-01-01')"
        )
    }}
),

final as (
    select
        cast(date_day as date) as date_day
    from base_dates
)

select *
from final
where date_day > dateadd(year, -5, current_date())  -- Keep recent dates only
  and date_day < dateadd(day, 30, current_date());
```
</File>

This generates a table of daily dates ranging from 5 years in the past to 30 days into the future.

## Add YAML configuration for the time spine

Now that you've created the SQL file, configure it in YAML so MetricFlow can recognize and use it.

1. Navigate to the `models/marts` directory.
2. Add a new YAML file named `_models.yml` with the following content:

<File name='models/marts/_models.yml'>

```yaml
models:
  - name: time_spine_daily
    description: A time spine with one row per day, ranging from 5 years in the past to 30 days into the future.
    time_spine:
      standard_granularity_column: date_day  # The base column used for time joins
    columns:
      - name: date_day
        description: The base date column for daily granularity
        granularity: day
```
</File>

This time spine YAML file:
- Defines `date_day` as the base column for daily granularity.

> - Adds [tests](/docs/build/data-tests) for uniqueness and non-null values.

Author: ```suggestion
```

- Configures `time_spine` properties so MetricFlow can use the table

### Using an existing dim_date table

This optional approach reuses an existing table, saving you the effort of creating a new one. However if you created a time spine table from scratch, you can skip this section. 

If your project already includes a `dim_date` or similar table, you can configure it as a time spine:

1. Locate the existing table (`dim_date`).
2. Update `_models.yml` file to configure it as a time spine:

<File name='_models.yml'>

```yaml
models:
  - name: dim_date
    description: An existing date dimension table used as a time spine.
    time_spine:
      standard_granularity_column: date_day
    columns:
      - name: date_day
        granularity: day
```
</File>

This time spine YAML file configures the `time_spine` property so MetricFlow can use the table.

## Run and test the time spine

> For the time spine table you created, let's run it and validate the output.

Author: @patkearns10 , i think maybe lines 133-143 dont make sense. i probably should just start with 144, query `dbt sl query --metrics revenue --group-by metric_time` -- does that make sense to you in terms of how to test the .sql and yaml file work?
![Screenshot 2025-02-07 at 17 23 16](https://github.com/user-attachments/assets/fdda7088-8f6a-4de7-b16a-803a3dd5424e)

Reviewer: eh, I can imagine most people are building the time spine for the first time. And it will fail if they do not run that time spine model.
Why do you want to not include those lines?

Author: i think it's because when i ran dbt run, i didn't see anything that confirmed it for me -- maybe in just the compiled code?

Author: ❓ confirm this 

oh i think i need to clarify that they need to preview the sql file (in a cli it'll be dbt show)

![Screenshot 2025-02-27 at 11 04 25](https://github.com/user-attachments/assets/3e43f0ca-3407-41aa-a7b0-6c549ef14ecb)

- Create a time spine SQL table from scratch, or
- Use an existing table in your project, like a `dim_date` table

And once you have a time spine table, you need to configure it in YAML to tell MetricFlow how to use it. This guide will show you how to do both!

## Add a time spine SQL model

Let's get started by assuming you're creating a time spine table from scratch. If you have a dbt project set up already and have your own time spine table (like a `dim_date` type model), you can skip this step and go to [Use an existing dim_date table](https://docs.getdbt.com//guides/mf-time-spine#using-an-existing-dim-date-table).

The time spine table is a dbt model that generates a series of dates (or timestamps) at a specific granularity. In this example, let's create a daily time spine table &mdash; `time_spine_daily.sql`. 

1. Navigate to the `models/marts` directory in your dbt project.
2. Add a new SQL file named `time_spine_daily.sql` with the following content:

<File name='models/marts/time_spine_daily.sql'>

```sql
{{
    config(
        materialized = 'table',
        schema = 'marts',

>         alias = 'time_spine_daily'

Reviewer: suggestion: remove these two congigs, as it is not required for this

    )
}}

with

base_dates as (
    {{
        dbt.date_spine(
            'day',
            "DATE('2000-01-01')",
            "DATE('2030-01-01')"
        )
    }}
),

final as (
    select
        cast(date_day as date) as date_day
    from base_dates
)

select *
from final
where date_day > dateadd(year, -5, current_date())  -- Keep recent dates only
  and date_day < dateadd(day, 30, current_date());
```
</File>

This generates a table of daily dates ranging from 5 years in the past to 30 days into the future.

## Add YAML configuration for the time spine

Now that you've created the SQL file, configure it in YAML so MetricFlow can recognize and use it.

1. Navigate to the `models/marts` directory.
2. Add a new YAML file named `_models.yml` with the following content:

<File name='models/marts/_models.yml'>

```yaml
models:
  - name: time_spine_daily

>     description: A time spine with one row per day, ranging from 2000-01-01 to 2030-01-01.

Reviewer: ```suggestion
    description: A time spine with one row per day, ranging from 5 years in the past to 30 days into the future.
```

    time_spine:
      standard_granularity_column: date_day  # The base column used for time joins
    columns:
      - name: date_day
        description: The base date column for daily granularity
        tests:
          - not_null

>           - unique

Reviewer: suggestion: get rid of this as it is not required

4. Double check that the results are correct and returning the expected data.

## Add additional granularities (optional)

To support multiple granularities (like hourly, monthly), create additional time spine tables and configure them in YAML.

1. Add a new SQL file named `time_spine_hourly.sql` with the following content:
<File name='models/marts/time_spine_hourly.sql'>

```sql
{{
    config(
        materialized = 'table',
    )
}}

with hours as (

    {{
        dbt.date_spine(
            'hour',
            "to_date('01/01/2000','mm/dd/yyyy')",
            "to_date('01/01/2025','mm/dd/yyyy')"
        )
    }}

),

final as (
    select cast(date_hour as timestamp) as date_hour
    from hours
)

select * from final
-- filter the time spine to a specific range
where date_hour > dateadd(year, -4, current_timestamp()) 
  and date_hour < dateadd(day, 30, current_timestamp())

```
</File>

2. Then update `_models.yml` file and add the hourly time spine (below the daily time spine config):

<File name='_models.yml'>

```yaml
models:
  - name: time_spine_daily
    ... rest of the daily time spine config ...

- name: time_spine_hourly

>     description: A time spine with one row per hour, ranging from 2000-01-01 to 2030-01-01.

Reviewer: ```suggestion
    description: A time spine with one row per hour, ranging from four years ago to today.
```

## Run and test the time spine

For the time spine table you created, let's run it and validate the output.

1. Run the time spine model to create the table:
   ```bash
   dbt run --select time_spine_daily
   ```

2. Validate the output by querying the generated table:
   ```sql
   select * from your_project_schema.time_spine_daily
   ```

3. Check that the table:
   - Contains one row per day.
   - Covers the date range you want (5 years back to 30 days forward)

4. (Optional) If you have [metrics](/docs/build/metrics) already defined in your project, you can query the table/metrics using [Semantic Layer commands](/docs/build/metricflow-commands) to validate the time spine. 
   
   Let's say you have a `revenue` metric defined. You can query the table/metrics using the following command:

    ```bash
    dbt sl query --metrics revenue --group-by metric_time
    ```

    This will output results similar to the following in the dbt Cloud IDE:
    <Lightbox src="/img/quickstarts/dbt-cloud/validate-mf-timespine-output.jpg" title="Validate the time spine output in dbt Cloud IDE" />

5. Double check that the results are correct and returning the expected data.

> ## Add additional granularities (optional)

Author: i thought we could do this based on [this example](https://docs.getdbt.com/docs/build/metricflow-time-spine#creating-a-time-spine-table) but when i tested it, it only returned date_day.

Reviewer: did you get this working? I would be happy to hop on a call and co-dev-test together if you can find overlapping time!

```sql
{{
    config(
        materialized = 'table',
    )
}}

with hours as (

    {{
        dbt.date_spine(
            'hour',
            "to_date('01/01/2000','mm/dd/yyyy')",
            "to_date('01/01/2025','mm/dd/yyyy')"
        )
    }}

),

final as (
    select cast(date_hour as timestamp) as date_hour
    from hours
)

select * from final
-- filter the time spine to a specific range
where date_hour > dateadd(year, -4, current_timestamp()) 
  and date_hour < dateadd(day, 30, current_timestamp())

```
</File>

2. Then update `_models.yml` file and add the hourly time spine (below the daily time spine config):

<File name='_models.yml'>

```yaml
models:
  - name: time_spine_daily
    ... rest of the daily time spine config ...

- name: time_spine_hourly
    description: A time spine with one row per hour, ranging from 2000-01-01 to 2030-01-01.
    time_spine:
      standard_granularity_column: date_hour
    columns:
      - name: date_hour
        granularity: hour
```
</File>

3. Run the model to create the table:
   ```bash
   dbt run --select time_spine_hourly
   ```

4. Validate the output by querying the generated table:
   ```bash

>    dbt sl query --metrics revenue --group-by metric_time__hour

Author: also the ide doesn't support `metric_time__hour` -- is this right?

Author: or should i scrap this section?

Reviewer: I would think the IDE supports `metric_time__hour`, why do you say otherwise?

You could try 
`dbt sl list dimensions --metrics revenue`

1. Navigate to the `models/marts` directory in your dbt project.
2. Add a new SQL file named `time_spine_daily.sql` with the following content:

    <File name='models/marts/time_spine_daily.sql'>

    ```sql
    {{
        config(
            materialized = 'table',
        )
    }}

    with

    base_dates as (
        {{
            dbt.date_spine(
                'day',
                "DATE('2000-01-01')",
                "DATE('2030-01-01')"
            )
        }}
    ),

    final as (
        select
            cast(date_day as date) as date_day
        from base_dates
    )

    select *
    from final
    where date_day > dateadd(year, -5, current_date())  -- Keep recent dates only
      and date_day < dateadd(day, 30, current_date());
    ```
    </File>
      
      This generates a table of daily dates ranging from 5 years in the past to 30 days into the future.

3. Run and preview the model to create the table:
    ```bash
    dbt run --select time_spine_daily 
    dbt show --select time_spine_daily # Use this command to preview the table if developing locally
    ```

> 4. If developing in the dbt Cloud IDE, you can preview the table by clicking the **Preview** button in the model page:

Author: ```suggestion
4. If developing in the dbt Cloud IDE, you can preview the table by clicking the **Preview** button:
```

### Using an existing dim_date table

This optional approach reuses an existing table, saving you the effort of creating a new one. However if you created a time spine table from scratch, you can skip this section. 

If your project already includes a `dim_date` or similar table, you can configure it as a time spine:

1. Locate the existing table (`dim_date`).
2. Update `_models.yml` file to configure it as a time spine:

    <File name='_models.yml'>

    ```yaml
    models:
      - name: dim_date
        description: An existing date dimension table used as a time spine.
        time_spine:
          standard_granularity_column: date_day
        columns:
          - name: date_day
            granularity: day
    ```
    </File>

    This time spine YAML file configures the `time_spine` property so MetricFlow can use the table.

## Run and preview the time spine

For the time spine table you created, let's run it and preview the output.

1. Run the following command:
   ```bash
   dbt run --select time_spine_daily
   dbt show --select time_spine_daily # Use this command to preview the table if developing locally
   ```

> 2. If developing in the dbt Cloud IDE, you can preview the table by clicking the **Preview** button in the model page:

Author: ```suggestion
2. If developing in the dbt Cloud IDE, you can preview the table by clicking the **Preview** button:
```

        }}

    ),

    final as (
        select cast(date_year as date) as date_year
        from years
    )

    select * from final
    -- filter the time spine to a specific range
    where date_year >= date_trunc('year', dateadd(year, -4, current_timestamp())) 
      and date_year < date_trunc('year', dateadd(year, 1, current_timestamp()))

    ```
    </File>

2. Then update `_models.yml` file and add the yearly time spine (below the daily time spine config):
    <File name='_models.yml'>

    ```yaml
    models:
      - name: time_spine_daily
        ... rest of the daily time spine config ...

    - name: time_spine_yearly
        description: time spine one row per house
        time_spine:
          standard_granularity_column: date_year
        columns:
          - name: date_year
            granularity: year
    ```
    </File>

3. Run or preview the model to create the table:
   ```bash
   dbt run --select time_spine_yearly
   dbt show --select time_spine_yearly # Use this command to preview the table if developing locally
   ```

4. Validate the output by querying the generated table:
   ```bash
   dbt sl query --metrics orders --group-by metric_time__year
   ```

> 

Author: ```suggestion

If you're developing in the dbt Cloud IDE, you can preview the table by clicking the **Preview** button.
```

    )

    select * from fiscal_calendar
    ```

    </File>

2. Then update `_models.yml` file and add the fiscal calendar time spine (below the yearly time spine config):
    <File name='_models.yml'>

    ```yaml
    models:
      - name: time_spine_yearly
        ... rest of the yearly time spine config ...  
        
      - name: fiscal_calendar
        description: A custom fiscal calendar with fiscal year and fiscal week granularities.
        time_spine:
          standard_granularity_column: date_day
          custom_granularities:
            - name: fiscal_year
              column_name: fiscal_year
            - name: fiscal_week
              column_name: fiscal_week
        columns:
          - name: date_day
            granularity: day
          - name: fiscal_year
            description: "Custom fiscal year starting in October"
          - name: fiscal_week
            description: "Fiscal week, shifted by 1 week from standard calendar"
    ```
    </File>

3. Run or preview the model to create the table:
   ```bash
   dbt run --select fiscal_calendar
   dbt show --select fiscal_calendar # Use this command to preview the table if developing locally
   ```

>    If you're developing in the dbt Cloud IDE, you can preview the table by clicking the **Preview** button in the model page.

Author: ```suggestion
   If you're developing in the dbt Cloud IDE, you can preview the table by clicking the **Preview** button.
```

---
title: "Quickstart with MetricFlow time spine"
id: "mf-time-spine"
level: 'Intermediate'
icon: 'guides'
tags: ['Quickstart', 'Semantic Layer']
hide_table_of_contents: true
recently_updated: true
---

<div style={{maxWidth: '900px'}}>

> ## Introduction

Author: change table to model (time spine table or view)

This guide explains how to configure a time spine using the [dbt Semantic Layer Jaffle shop example project](https://github.com/dbt-labs/jaffle-sl-template) as a reference. 

> ### What is a time spine table?

Author: ```suggestion
### What is a time spine model?
```

        time_spine:
          standard_granularity_column: date_day  # The base column used for time joins
        columns:
          - name: date_day
            description: The base date column for daily granularity
            granularity: day
    ```
    </File>

This time spine YAML file:
- Defines `date_day` as the base column for daily granularity.
- Configures `time_spine` properties so MetricFlow can use the table

### Using an existing dim_date table

This optional approach reuses an existing table, saving you the effort of creating a new one. However if you created a time spine table from scratch, you can skip this section. 

If your project already includes a `dim_date` or similar table, you can configure it as a time spine:

1. Locate the existing table (`dim_date`).
2. Update `_models.yml` file to configure it as a time spine:

    <File name='_models.yml'>

    ```yaml
    models:
      - name: dim_date
        description: An existing date dimension table used as a time spine.
        time_spine:
          standard_granularity_column: date_day
        columns:

>           - name: date_day

Author: add a couple of more columns to show other stuff. use pat's example

### Yearly time spine

To support multiple granularities (like hourly, yearly, monthly), create additional time spine tables and configure them in YAML.

1. Add a new SQL file named `time_spine_yearly.sql` with the following content:
    <File name='models/marts/time_spine_yearly.sql'>

    ```sql
    {{
        config(
            materialized = 'table',
        )
    }}

    with years as (

        {{
            dbt.date_spine(
                'year',
                "to_date('01/01/2000','mm/dd/yyyy')",
                "to_date('01/01/2025','mm/dd/yyyy')"
            )
        }}

    ),

    final as (
        select cast(date_year as date) as date_year
        from years
    )

    select * from final
    -- filter the time spine to a specific range
    where date_year >= date_trunc('year', dateadd(year, -4, current_timestamp())) 
      and date_year < date_trunc('year', dateadd(year, 1, current_timestamp()))

    ```
    </File>

2. Then update `_models.yml` file and add the yearly time spine (below the daily time spine config):
    <File name='_models.yml'>

    ```yaml
    models:
      - name: time_spine_daily
        ... rest of the daily time spine config ...

>     - name: time_spine_yearly

Author: move indentation

        description: time spine one row per house
        time_spine:
          standard_granularity_column: date_year
        columns:
          - name: date_year
            granularity: year
    ```
    </File>

3. Run or preview the model to create the table:
   ```bash
   dbt run --select time_spine_yearly
   dbt show --select time_spine_yearly # Use this command to preview the table if developing locally
   ```

4. Validate the output by querying the generated table:
   ```bash
   dbt sl query --metrics orders --group-by metric_time__year
   ```

If you're developing in the dbt Cloud IDE, you can preview the table by clicking the **Preview** button.
   <Lightbox src="/img/mf-guide-query.jpg" title="Validate the metrics and time spine output in dbt Cloud IDE" />

:::tip Extra credit!
For some extra practice, try one of the following exercises:

- Order the `dbt sl query --metrics orders --group-by metric_time__year` command output by ascending order of `metric_time__year`. Check out the [dbt Semantic Layer commands](/docs/build/metricflow-commands#query) docs for more information on how to do this.

> - Try creating a monthly time spine &mdash; duplicate your daily time spine model, adjust it to generate one row per month, and update the YAML file to include `granularity: month`. Give it a try!

Author: extra bullet -- filter to this year and last year only to limit data returned

    dbt show --select time_spine_daily # Use this command to preview the model if developing locally
    ```

4. If developing in the dbt Cloud IDE, you can preview the model by clicking the **Preview** button:
   <Lightbox src="/img/mf-guide-preview-time-spine-table.jpg" title="Preview the time spine model in dbt Cloud IDE" />

## Add YAML configuration for the time spine

Now that you've created the SQL file, configure it in YAML so MetricFlow can recognize and use it.

1. Navigate to the `models/marts` directory.
2. Add a new YAML file named `_models.yml` with the following content:

    <File name='models/marts/_models.yml'>

    ```yaml
    models:
      - name: time_spine_daily
        description: A time spine with one row per day, ranging from 5 years in the past to 30 days into the future.
        time_spine:
          standard_granularity_column: date_day  # The base column used for time joins
        columns:
          - name: date_day
            description: The base date column for daily granularity
            granularity: day
    ```
    </File>

This time spine YAML file:
- Defines `date_day` as the base column for daily granularity.

> - Configures `time_spine` properties so MetricFlow can use the model

Reviewer 2: ```suggestion
- Configures `time_spine` properties so MetricFlow can use the model.
```

    <File name='_models.yml'>

    ```yaml
    models:
      - name: dim_date
        description: An existing date dimension model used as a time spine.
        time_spine:
          standard_granularity_column: date_day
        columns:
          - name: date_day
            granularity: day
          - name: day_of_week
            granularity: day
          - name: full_date
            granularity: day
    ```
    </File>

    This time spine YAML file configures the `time_spine` property so MetricFlow can use the model.

## Run and preview the time spine

For the time spine you created, let's run it and preview the output if you haven't already done this. If you have already run the model, you can skip this step.

1. Run the following command:
   ```bash
   dbt run --select time_spine_daily
   dbt show --select time_spine_daily # Use this command to preview the model if developing locally
   ```

2. If developing in the dbt Cloud IDE, you can preview the model by clicking the **Preview** button:
    <Lightbox src="/img/mf-guide-preview-time-spine-table.jpg" title="Preview the time spine model in dbt Cloud IDE" />

3. Check that the model:
   - Contains one row per day.

>    - Covers the date range you want (5 years back to 30 days forward)

Reviewer 2: ```suggestion
   - Covers the date range you want (5 years back to 30 days forward).
```

## Add additional granularities

This section is optional and will show you how to add additional granularities to your time spine:

- [Yearly](#yearly-time-spine)
- [Custom calendars](#custom-calendars)

### Yearly time spine

To support multiple granularities (like hourly, yearly, monthly), create additional time spine models and configure them in YAML.

1. Add a new SQL file named `time_spine_yearly.sql` with the following content:
    <File name='models/marts/time_spine_yearly.sql'>

    ```sql
    {{
        config(
            materialized = 'table',
        )
    }}

    with years as (

        {{
            dbt.date_spine(
                'year',
                "to_date('01/01/2000','mm/dd/yyyy')",
                "to_date('01/01/2025','mm/dd/yyyy')"
            )
        }}

    ),

    final as (
        select cast(date_year as date) as date_year
        from years
    )

    select * from final
    -- filter the time spine to a specific range
    where date_year >= date_trunc('year', dateadd(year, -4, current_timestamp())) 
      and date_year < date_trunc('year', dateadd(year, 1, current_timestamp()))

    ```
    </File>

> 2. Then update `_models.yml` file and add the yearly time spine (below the daily time spine config):

Reviewer 2: ```suggestion
2. Then update the `_models.yml` file and add the yearly time spine (below the daily time spine config):
```
